provider/equinix/environ_network.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package equinix
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/names/v4"
+	"github.com/juju/utils/set"
+	"github.com/packethost/packngo"
+)
+
+var _ environs.Networking = (*environ)(nil)
+
+// Subnets returns basic information about subnets known by the provider for
+// the environment.
+func (e *environ) Subnets(ctx context.ProviderCallContext, inst instance.Id, subnetIDs []network.Id) ([]network.SubnetInfo, error) {
+	attrs := e.cloud.Credential.Attributes()
+	if attrs == nil {
+		return nil, errors.Trace(fmt.Errorf("empty attribute credentials"))
+	}
+	// We checked the presence of project-id when we were verifying the credentials.
+	projectID := attrs["project-id"]
+	ips, _, err := e.equinixClient.ProjectIPs.List(projectID, &packngo.ListOptions{
+		Includes: []string{"available_in_metros"},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var projectSubnets []network.SubnetInfo
+	filterSet := set.NewStrings()
+	for _, ipblock := range ips {
+		subnetID, cidr, err := makeSubnetIDForNetwork(ipblock.ID, ipblock.Network, ipblock.CIDR)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		for _, i := range subnetIDs {
+			filterSet.Add(i.String())
+		}
+
+		if ipblock.Facility.Metro == nil || ipblock.Facility.Metro.Code != e.cloud.Region || (!filterSet.IsEmpty() && !filterSet.Contains(subnetID)) {
```

Comment:
ipblock.Facility is a pointer and can be nil. In fact, if you try bootstrapping as-is to the 'am' region you will get a panic here.

Reply:
I asked @displague to have a look at this because for some reason `ipblock.Metro` and `ipblock.Facility.Metro` both exists. And it sounds not right